pkg/controllers/disruption/controller.go

```diff
@@ -179,194 +163,60 @@ func (c *Controller) disrupt(ctx context.Context, disruption Method) (bool, erro
 	return true, nil
 }
 
+// executeCommand will do the following, untainting if the step fails.
+// 1. Taint candidate nodes
+// 2. Spin up replacement nodes
+// 3. Add Command to orchestration.Queue to wait to delete the candiates.
 func (c *Controller) executeCommand(ctx context.Context, m Method, cmd Command) error {
-	deprovisioningActionsPerformedCounter.With(map[string]string{
-		actionLabel:        fmt.Sprintf("%s/%s", m.Type(), cmd.Action()),
-		deprovisionerLabel: m.Type(),
-	}).Inc()
 	disruptionActionsPerformedCounter.With(map[string]string{
 		actionLabel:            string(cmd.Action()),
 		methodLabel:            m.Type(),
 		consolidationTypeLabel: m.ConsolidationType(),
 	}).Inc()
 	logging.FromContext(ctx).Infof("disrupting via %s %s", m.Type(), cmd)
 
-	reason := fmt.Sprintf("%s/%s", m.Type(), cmd.Action())
-	if cmd.Action() == ReplaceAction {
-		if err := c.launchReplacementNodeClaims(ctx, m, cmd); err != nil {
-			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
-			// we don't want to disrupt workloads with no way to provision new NodeClaims for them.
-			return fmt.Errorf("launching replacement, %w", err)
-		}
+	stateNodes := lo.Map(cmd.candidates, func(c *Candidate, _ int) *state.StateNode {
+		return c.StateNode
+	})
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := state.RequireNoScheduleTaint(ctx, c.kubeClient, true, stateNodes...); err != nil {
+		return multierr.Append(fmt.Errorf("tainting nodes, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
 	}
 
-	for _, candidate := range cmd.candidates {
-		c.recorder.Publish(disruptionevents.Terminating(candidate.Node, candidate.NodeClaim, reason)...)
-
-		if err := nodeclaimutil.Delete(ctx, c.kubeClient, candidate.NodeClaim); err != nil {
-			if !errors.IsNotFound(err) {
-				logging.FromContext(ctx).Errorf("terminating, %s", err)
-			}
-			continue
+	var nodeClaimKeys []nodeclaim.Key
+	var err error
+	if len(cmd.replacements) > 0 {
+		if nodeClaimKeys, err = c.createReplacementNodeClaims(ctx, m, cmd); err != nil {
+			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			return multierr.Append(fmt.Errorf("launching replacement nodeclaim, %w", err), state.RequireNoScheduleTaint(ctx, c.kubeClient, false, stateNodes...))
```

Comment:
This error chaining logic seems wrong. What if `state.RequireNoScheduleTaint` doesn't error? Then you're returning an error that wraps a non-error.

What if `state.RequireNoScheduleTaint` does fail. Does it then contain the taint forever?

Reply:
This is a really good point that I wasn't thinking much about until now since this is the way that it's been done up to this point and this already existed as a failure mode, but we should think about how to make the taint add stateless. Can we tie the association of the taint to the node with its presence in the queue in some way so that removing the taint is retryable?

Reply:
In the recent past, this was a bit worse, where we wouldn't uncordon/untaint nodes at the start of the disruption loop. Nodes should be untainted if this fails since we retry at the start of each disruption reconciliation, where this PR links the candidate's presence in the queue to if it should have the taint or not. 

On the error chaining point, `multierr.Append()` will just return the non-nil error if one of the errors is nil. So in this case, it would read:
`launching replacement nodelclaim, command timed out, failed to untaint nodes` or just `launching replacement nodelclaim, command timed out`.